src/main/java/org/montrealjug/billetterie/ui/BookersController.java

```diff
@@ -0,0 +1,141 @@
+
+        bookers.forEach(
+                (booker -> {
+                    presentationBookers.add(
+                            new PresentationBooker(
+                                    booker.getFirstName(),
+                                    booker.getLastName(),
+                                    booker.getEmail()));
+                }));
+
+        model.addAttribute("bookerList", presentationBookers);
+
+        return "bookers-list";
+    }
+
+    @PostMapping("")
+    public ResponseEntity<Void> addBooker(@Valid PresentationBooker booker, Model model)
+            throws Exception {
+        Booker bookerEntity = new Booker();
+
+        String email = booker.email().toLowerCase();
+
+        if (bookerRepository.existsById(email)) {
+            throw new RuntimeException("User with this email already exists");
+        }
+
+        bookerEntity.setEmail(email);
+        bookerEntity.setFirstName(booker.firstName());
+        bookerEntity.setLastName(booker.lastName());
+        bookerEntity.setValidationTime(Instant.now());
+        bookerEntity.setEmailSignature(signatureService.signAndTrim(email));
+
+        bookerRepository.save(bookerEntity);
+
+        return ResponseEntity.status(HttpStatus.FOUND)
+                .location(URI.create("/admin/bookers"))
+                .build();
+    }
+
+    @GetMapping("add-booker")
+    public String showAddBookerPage() {
+        return "bookers-create-update";
+    }
+
+    @GetMapping("{email}")
+    public String showUpdateBookerPage(Model model, @PathVariable String email) {
+        Optional<Booker> optionalBooker = this.bookerRepository.findById(email);
+
+        if (optionalBooker.isEmpty()) {
+            throw new EntityNotFoundException(
+                    "Booker with email " + email + " not found", "bookers-create-update");
+        }
+
+        Booker booker = optionalBooker.get();
+        PresentationBooker presentationBooker =
+                new PresentationBooker(
+                        booker.getFirstName(), booker.getLastName(), booker.getEmail());
+        model.addAttribute("booker", presentationBooker);
+
+        return "bookers-create-update";
+    }
+
+    @PostMapping("{email}")
+    public ResponseEntity<Void> updateBooker(
+            Model model, @PathVariable String email, @Valid PresentationBooker presentationBooker)
+            throws Exception {
+        Optional<Booker> optionalBooker = bookerRepository.findById(email);
+
+        if (optionalBooker.isEmpty()) {
+            throw new RedirectableNotFoundException(
+                    "Booker with email " + email + " not found", "/admin/bookers/" + email);
+        }
+
+        Booker booker = optionalBooker.get();
+
+        booker.setFirstName(presentationBooker.firstName());
+        booker.setLastName(presentationBooker.lastName());
+
+        //         TODO allow changing user's email address (currently unable since email is the
```

Comment:
ah, fair point.

that's probably fine like this: if the user wants to change their email, they'll have to register a new email, associate new participants and probably delete the old email / booker

we'll see how that goes...

Reply:
see https://github.com/montrealjug/billetterie/discussions/33